### PR TITLE
Excel Tests Continued Fixture Cleanup

### PR DESCRIPTION
**Pull request by WillAyd**
Follow up to #26579

Removing the frame2 fixture which is sparsely used. I've also removed inheritance of `ReadingTestsBase` and made that it's own discoverable class. Previously these tests were only being generated when `TestXlrdReader` was being run which tightly couples reading with xlrd. 

The goal here in subsequent PRs would be to better parametrize `TestReaders` for new engines that come in and truly make `TestXlrdReader` only deal with xlrd specific tests (will have to move around some tests to make that happen)

### CODE REVIEW

File: pandas/tests/io/test_excel.py [hunk 1]
```diff
@@ -58,8 +51,9 @@ def ignore_xlrd_time_clock_warning():
         yield
 
 
-@td.skip_if_no('xlrd', '1.0.0')
-class ReadingTestsBase:
+@td.skip_if_no('xlrd')
```
Comment (WillAyd) [inline pandas/tests/io/test_excel.py]: Note that our project-wide minimum ix 1.0.0, so this decorator was being duplicative.

This decorator also won't be there in the long term once we completely couple TestReaders from xlrd, but keeping here for now

**Comment by WillAyd** [comment]
This was causing failures for me locally per this convo:

https://github.com/pandas-dev/pandas/pull/26657#issuecomment-499081131

Let's see what happens here...

File: pandas/tests/io/test_excel.py [hunk 2]
```diff
@@ -58,11 +51,16 @@ def ignore_xlrd_time_clock_warning():
...
-class ReadingTestsBase:
-    # This is based on ExcelWriterBase
-
-    @pytest.fixture(autouse=True, params=['xlrd', None])
+@pytest.mark.parametrize("ext", ['.xls', '.xlsx', '.xlsm'])
+class TestReaders:
+
+    @pytest.fixture(autouse=True, params=[
+        # Add any engines to test here
+        pytest.param('xlrd', marks=pytest.mark.skipif(
```
Comment (WillAyd) [inline pandas/tests/io/test_excel.py]: Might be a cleaner way to mark these but in any case this is more or less what I see being easiest long term as we look at add openpyxl into the reading mix. Any reader can be parametrized here with this fixture and automatically cover all of the reading tests.

I'll have to move or split some tests out of the base class here that truly do test xldd-specific behavior but leaving that for another follow up

File: pandas/tests/io/test_excel.py [hunk 3]
```diff
@@ -80,7 +83,6 @@ def df_ref(self):
                           parse_dates=True, engine='python')
         return df_ref
 
-    @td.skip_if_no("xlrd", "1.0.1")  # see gh-22682
```
Comment (WillAyd) [inline pandas/tests/io/test_excel.py]: These didn't really solve the issue they were supposed to so removed to make things cleaner. Again I think the real issue occurs when defusedxml is installed in the right environment, but will have to delve further into which environments that is exactly (orthogonal to this PR)

File: pandas/tests/io/test_excel.py [hunk 4]
```diff
@@ -844,7 +841,7 @@ def test_read_excel_squeeze(self, ext):
         tm.assert_series_equal(actual, expected)
 
 
-@td.skip_if_no('xlrd', '1.0.0')
+@td.skip_if_no('xlrd')
```
Comment (WillAyd) [inline pandas/tests/io/test_excel.py]: Comment might have been lost on a prior PR but the min xlrd version for pandas is 1.0.0 anyway, so this was duplicative

File: pandas/tests/io/test_excel.py [hunk 5]
```diff
@@ -1135,7 +1131,7 @@ def test_excel_sheet_size(self):
         with pytest.raises(ValueError, match=msg):
             col_df.to_excel(self.path)
 
-    def test_excel_sheet_by_name_raise(self, *_):
+    def test_excel_sheet_by_name_raise(self, engine, ext):
```
Comment (WillAyd) [inline pandas/tests/io/test_excel.py]: The `*_` signature was making practically all fixtures available to the function, so replaced with something explicit (easier now that we've decoupled the class based designed from the fixturized design)

File: pandas/tests/io/test_excel.py [hunk 6]
```diff
@@ -32,18 +32,16 @@ def frame(float_frame):
...
-    return float_frame[:10]
-
-
 @pytest.fixture
 def tsframe():
     return tm.makeTimeDataFrame()[:5]
 
 
+@pytest.fixture(params=[True, False])
+def merge_cells(request):
```
Comment (WillAyd) [inline pandas/tests/io/test_excel.py]: This was previously a decorator for the writing tests but used in less than half of those tests, so it was adding unnecessary parametrization. By decoupling this from the class we can simply inject where necessary

**Review by WillAyd — COMMENTED** [review]
OK think I'm stopping here. This should be a pretty good cleanup to remove unnecessary fixturation.

I think after this going to align the engine / ext usage in the Writer classes and can start splitting into sub-modules thereafter

**Comment by jreback** [comment]
lgtm. @simonjayhawkins

**Review by simonjayhawkins — COMMENTED** [review]
@WillAyd generally lgtm. a few suggestions.

File: pandas/tests/io/test_excel.py (hunk 2 above)
Comment (WillAyd) [reply pandas/tests/io/test_excel.py]: That's an option but once we start adding more engines with different file extensions (ex: openpyxl won't handle xlrd, odfpy won't be excel files) that would generate a ton of skips. One option to reduce that would be to change the scope here to a higher level, but I think it makes sense to defer that or whatever the ultimate change is for now

File: pandas/tests/io/test_excel.py [hunk 7]
```diff
@@ -1045,9 +1042,9 @@ def test_read_excel_parse_dates(self, ext):
             tm.assert_frame_equal(df, res)
 
 
-@td.skip_if_no('xlrd', '1.0.0')
+@td.skip_if_no('xlrd')
 @pytest.mark.parametrize("ext", ['.xls', '.xlsx', '.xlsm'])
-class TestXlrdReader(ReadingTestsBase):
+class TestXlrdReader:
```
[the remark this answers is not in the document]
Comment (WillAyd) [reply pandas/tests/io/test_excel.py]: Yes that is intentional. The purpose of this class is to just test the xlrd function, so bundling None in here somewhat blurs the intention.

Generally we could probably get rid of the `None` engine tests and just have a separate test that ensures how that gets bound to a particular engine, but deferring that as well for now

File: pandas/tests/io/test_excel.py (hunk 5 above)
Comment (WillAyd) [reply pandas/tests/io/test_excel.py]: > engine and ext do not need to be in the function signature if their values are not required by the test. the parameterisation is handled by the set_engine_and_path autouse fixture.

Good to know. Should definitely be able to simplify some of these methods on next push then!

File: pandas/tests/io/test_excel.py (hunk 6 above)
Comment (WillAyd) [reply pandas/tests/io/test_excel.py]: No problem - will do on next push

File: pandas/tests/io/test_excel.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (simonjayhawkins) [reply pandas/tests/io/test_excel.py]: makes sense.

File: pandas/tests/io/test_excel.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/tests/io/test_excel.py]: ok, for now then as will also depend on how engine=None will be tested if that's changed.

**Comment by simonjayhawkins** [comment]
@WillAyd changes look good.

test failure is a mystery to me at the moment. I would say it was unrelated if it wasn't for the fact that no other builds are seeing this.

the failing test has an engine parameter. coincidence? no reason that changes to test_excel should affect this test. test output shows `engine = 'numexpr'` so implies no clash of fixtures.

the parametrisation of the engine fixture in pandas\tests\computation\test_eval.py does look iffy though.

**Comment by WillAyd** [comment]
Yea no rush to merge I’ll take a look and maybe push master. Seems unlikely but a clash in fixture naming could be possible.

**Comment by simonjayhawkins** [comment]
i think the engine bit may be a coincidence.

it appears that a few tests are using set_use_numexpr which is messing with global variables instead of monkeypatching.

so if the order of tests change, could be any unrelated changes, this could cause a problem.

as an experiment, you could try changing the engine fixture in pandas\tests\computation\test_eval.py to

```
@pytest.fixture(params=_engines)
def engine(request):
    engine = request.param
    if engine == 'numexpr' and not _USE_NUMEXPR:
        pytest.skip('numexpr enabled->{enabled}, '
                    'installed->{installed}'.format(
                        enabled=_USE_NUMEXPR,
                        installed=_NUMEXPR_INSTALLED))
    return engine
```

so that the skip is evaluated at test execution instead of during test collection since skipping depends on the global variable.

**Comment by WillAyd** [comment]
Makes sense and I think what you have is more readable anyway. Just pushed

**Comment by simonjayhawkins** [comment]
i have a suspicion that'll probably not work because the test is quite involved so there will still be lag between checking the global and performing the failing op.

not sure how much time you want to spend on this. could just revert the last couple of commits for now.

i've had a go a parametrising the tests to split them up. but it may be easier to just debug the test on ci by adding print statements.

```
    def _test_basic_series_frame_alignment(
            self, engine, parser, r_idx_type, c_idx_type, index_name):
        with warnings.catch_warnings(record=True):
            # avoid warning about comparing strings and ints
            warnings.simplefilter("ignore", RuntimeWarning)
            df = mkdf(10, 7, data_gen_f=f, r_idx_type=r_idx_type,
                      c_idx_type=c_idx_type)
            index = getattr(df, index_name)
            s = Series(np.random.randn(5), index[:5])
            if should_warn(s.index, df.index):
                with tm.assert_produces_warning(RuntimeWarning):
                    res = pd.eval('s + df', engine=engine, parser=parser)
            else:
                res = pd.eval('s + df', engine=engine, parser=parser)

            if r_idx_type == 'dt' or c_idx_type == 'dt':
                expected = df.add(s) if engine == 'numexpr' else s + df
            else:
                expected = s + df
            assert_frame_equal(res, expected)

    @pytest.mark.parametrize('r_idx_type', ['i', 'u', 's'])
    @pytest.mark.parametrize('c_idx_type', ['i', 'u', 's'])
    @pytest.mark.parametrize('index_name', ['index', 'columns'])
    def test_basic_series_frame_alignment(
            self, engine, parser, r_idx_type, c_idx_type, index_name):
        self._test_basic_series_frame_alignment(
            engine, parser, r_idx_type, c_idx_type, index_name)

    @pytest.mark.parametrize('index_name', ['index', 'columns'])
    def test_basic_series_frame_alignment_dt(self, engine, parser, index_name):
        # only test dt with dt, otherwise weird joins result
        self._test_basic_series_frame_alignment(
            engine, parser, 'dt', 'dt', index_name)```

**Comment by WillAyd** [comment]
OK yea let me revert the last few changes then. Even if it passes CI somewhat fragile and orthogonal anyway. PR is big enough as is so happy to address separately and in coordination with whatever you are looking at

**Review by simonjayhawkins — APPROVED** [review]

**Comment by WillAyd** [comment]
Thanks for the review @simonjayhawkins

**Pull request merged**